windows/ReactNativeWebView/ReactWebViewManager.cpp

```diff
@@ -0,0 +1,267 @@
+                WriteWebViewNavigationEventArg(eventDataWriter);
+                eventDataWriter.WriteObjectEnd();
+            });
+    }
+
+    void ReactWebViewManager::OnNavigationCompleted(winrt::WebView const& webView, winrt::WebViewNavigationCompletedEventArgs const& /*args*/) {
+        m_reactContext.DispatchEvent(
+            webView,
+            L"topLoadingFinish",
+            [&](winrt::IJSValueWriter const& eventDataWriter) noexcept {
+                eventDataWriter.WriteObjectBegin();
+                WriteWebViewNavigationEventArg(eventDataWriter);
+                eventDataWriter.WriteObjectEnd();
+            });
+
+        winrt::hstring windowAlert = L"window.alert = function (msg) {window.external.notify(`{\"type\":\"alert\",\"message\":\"${msg}\"}`)};";
+        winrt::hstring postMessage = L"window.ReactNativeWebView = {postMessage: function (data) {window.external.notify(String(data))}};";
+        InjectJavaScript(windowAlert + postMessage);
+    }
+
+    void ReactWebViewManager::OnNavigationFailed(winrt::IInspectable const& /*sender*/, winrt::WebViewNavigationFailedEventArgs const& args) {
+        m_reactContext.DispatchEvent(
+            m_webView,
+            L"topLoadingError",
+            [&](winrt::IJSValueWriter const& eventDataWriter) noexcept {
+                auto httpCode = static_cast<int32_t>(args.WebErrorStatus());
+                eventDataWriter.WriteObjectBegin();
+                {
+                    WriteProperty(eventDataWriter, L"code", httpCode);
+                    WriteWebViewNavigationEventArg(eventDataWriter);
+                }
+                eventDataWriter.WriteObjectEnd();
+            });
+    }
+
+    void ReactWebViewManager::OnScriptNotify(winrt::IInspectable const& /*sender*/, winrt::Windows::UI::Xaml::Controls::NotifyEventArgs const& args) {
+        winrt::JsonObject jsonObject;
+        if (winrt::JsonObject::TryParse(args.Value(), jsonObject)) {
+            auto type = jsonObject.GetNamedString(L"type");
+            if (type == L"alert") {
+                auto dialog = winrt::MessageDialog(jsonObject.GetNamedString(L"message"));
+                dialog.Commands().Append(winrt::UICommand(L"OK"));
+                dialog.ShowAsync();
+            }
+        }
+        else {
+            m_reactContext.DispatchEvent(
+                m_webView,
+                L"topMessage",
+                [&](winrt::Microsoft::ReactNative::IJSValueWriter const& eventDataWriter) noexcept {
+                    eventDataWriter.WriteObjectBegin();
+                    {
+                        WriteProperty(eventDataWriter, L"data", winrt::to_string(args.Value()));
+                    }
+                    eventDataWriter.WriteObjectEnd();
+                });
+        }
+    }
+
+    winrt::IAsyncAction ReactWebViewManager::InjectJavaScript(winrt::hstring javaScript) {
+        auto result = co_await m_webView.InvokeScriptAsync(L"eval", { javaScript });

```

Comment:
Is there any way to do this without using eval?  Since Hermes apparently doesn't support it?